### ISSUE TITLE
Addon should be able to use sniff from external repo

### ISSUE THREAD

**Issue by ernilambar** [10up/plugin-check#740]
I am developing PCP addon which will use external sniff. Those are setup properly in the addon. But PHPCS runner does not detect such coding standard installed from the addon.

This is a minimal PCP addon for example: https://github.com/ernilambar/tiny-check

Coding standard I am trying to use: https://github.com/ernilambar/coding-standard/

`class Todo_Check extends Abstract_PHP_CodeSniffer_Check`

```php
return array(
	'extensions'  => 'php',
	'standard'    => 'NilambarCodingStandard',
	'sniffs'      => 'NilambarCodingStandard.Commenting.TodoComment',
);
```

This rule is not detected by our runner. Gives sniff not registered error.

For debugging, I did composer require this standard in PCP, and then this sniff is detected and check is working properly.

Is there a way to make addon's sniffs recognizable to the main PCP plugin?

**Comment by swissspidy** [comment]
It makes sense that `Abstract_PHP_CodeSniffer_Check` doesn't know about your custom sniff out of the box. There's no way for it to know where `NilambarCodingStandard ` it is coming from. How should it?

In PCP itself, when you install custom sniffs, what happens under the hood is that `phpcs --config-set installed_paths` is being called to add the location of those sniffs to `vendor/squizlabs/php_codesniffer/CodeSniffer.conf`. From there, PHPCS will add those paths to its autoloader (`vendor/squizlabs/php_codesniffer/autoload.php`).

What you could try:

1. Manually `require_once` your sniff class, so that it will be known by PHPCS.
2. Call `phpcs --config-set installed_paths` before and after the run (sounds very fragile)
3. Instead of using `Abstract_PHP_CodeSniffer_Check`, roll your own implementation

**Comment by ernilambar** [comment]
Yes, I understand. It's logical to not detect by itself. It would be pretty powerful feature if we are able to use external sniff in Plugin Check addon. Option 3 seems more interesting than other two. I will investigate more about it.

**Comment by ernilambar** [comment]
I created new abstract class based on `Abstract_PHP_CodeSniffer_Check`. Only thing I needed to make it work was changing autoload path in the new class.

```php
$autoloader = PCP_ADDON_DIR . '/vendor/squizlabs/php_codesniffer/autoload.php';
```

I dont want to maintain whole class in addon as it contains lots of stuffs. Any suggestion on overriding that autoloader path from the child class, like `set_autoloader_path( $full_path )`. Do you think it is feasible?

**Comment by swissspidy** [comment]
I'm not sure overriding the whole autoloader path is the right approach, as then you'd still have to maintain your own copy of PHPCS basically.

Have you tried ideas (1) and (2) already?

**Comment by ernilambar** [comment]
I tried approach 1. I included all PHP files manually using require_once:

```php
require_once PCP_ADDON_DIR . '/vendor/squizlabs/php_codesniffer/src/Sniffs/Sniff.php';
require_once PCP_ADDON_DIR . '/vendor/wp-coding-standards/wpcs/WordPress/Sniff.php';
require_once PCP_ADDON_DIR . '/vendor/wp-coding-standards/wpcs/WordPress/AbstractFunctionRestrictionsSniff.php';
require_once PCP_ADDON_DIR . '/vendor/wp-coding-standards/wpcs/WordPress/AbstractFunctionParameterSniff.php';
require_once PCP_ADDON_DIR . '/vendor/ernilambar/coding-standard/NilambarCodingStandard/Sniffs/Security/SettingSanitizationSniff.php';
```

Classes are loaded as expected but still I get `ERROR: Referenced sniff "NilambarCodingStandard.Security.SettingSanitization" does not exist` error. 

I will try approach 2 now.

**Comment by ernilambar** [comment]
I tried setting installed_paths using `runtime-set` argument like below but it did not work:

```php
'runtime-set' => array(
	'installed_paths' => PCP_ADDON_DIR . '/vendor/ernilambar/coding-standard',
),
```
I could not find the way to set and reset install_paths like you mentioned in approach 2. :-(

**Comment by swissspidy** [comment]
`config-set` can't be combined with the command to run the sniffs itself. It would need to be executed separately before and after.

I see now that (1) won't work because PHPCS looks in `installed_paths` first to see if the standard is installed.

So let me whip up a quick PR to demonstrate what I meant with (2)
